linkerd/app/outbound/src/endpoint.rs

```diff
@@ -54,6 +66,72 @@ impl<E> Outbound<E> {
             stack,
         }
     }
+
+    /// Pushes a layer that checks if a discovered service profile contains an
+    /// endpoint override, and forwards directly to that endpoint (bypassing the
+    /// current stack) if one exists.
+    pub fn push_endpoint_override<T, O, R>(
+        self,
+        ep_override: O,
+    ) -> Outbound<
+        impl svc::NewService<
+                (Option<profiles::Receiver>, T),
+                Service = OrOverride<E::Service, O::Service>,
+            > + Clone,
+    >
+    where
+        E: svc::NewService<(Option<profiles::Receiver>, T)> + Clone,
+        E::Service: svc::Service<R, Error = Error>,
+        <E::Service as svc::Service<R>>::Future: Send,
+        O: svc::NewService<ProfileOverride> + Clone,
+        O::Service:
+            svc::Service<R, Response = <E::Service as svc::Service<R>>::Response, Error = Error>,
+        <O::Service as svc::Service<R>>::Future: Send,
+        T: std::fmt::Debug,
+    {
+        let Self {
+            config,
+            runtime,
+            stack: no_override,
+        } = self;
+        let identity_disabled = runtime.identity.is_none();
+        let stack = no_override.push_switch(
+            move |(profile, target): (Option<profiles::Receiver>, T)| -> Result<_, Error> {
+                let rx = match profile {
+                    Some(profile) => profile,
+                    None => return Ok(svc::Either::A((None, target))),
+                };
+                {
+                let profile = rx.borrow();
+                if let Some((addr, metadata)) = profile.endpoint.clone() {
+                    tracing::debug!(%addr, ?metadata, ?target, "Using endpoint from profile override");
+                    let tls = if identity_disabled {
+                        tls::ConditionalClientTls::None(tls::NoClientTls::Disabled)
+                    } else {
+                        FromMetadata::client_tls(&metadata, tls::NoClientTls::NotProvidedByServiceDiscovery)
+                    };
+                    let logical_addr = profile.addr.clone().map(|LogicalAddr(addr)| Addr::from(addr)).unwrap_or_else(|| Addr::from(addr));
+                    let endpoint = Endpoint {
+                        addr: Remote(ServerAddr(addr)),
+                        tls,
+                        metadata,
+                        logical_addr,
+                        protocol: (),
+                    };
+                    let opaque_protocol = profile.opaque_protocol;
+                    return Ok(svc::Either::B(ProfileOverride { endpoint, opaque_protocol }));
+                }
+            };
```

Comment:
@olix0r as a follow-up, i'm thinking about also rolling this logic into this layer: https://github.com/linkerd/linkerd2-proxy/blob/94c7796351de6a7dc3ec6740e60a355719c37ecd/linkerd/app/outbound/src/logical.rs#L111-L114

wdyt? that way, all the "skip logical and go directly to the endpoint" logic is encapsulated in two places (this, and the "no profile" layer).